cli/cli_app.go

```diff
@@ -542,26 +542,44 @@ func runTerragruntWithConfig(terragruntOptions *options.TerragruntOptions, terra
 		return err
 	}
 
-	return runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() error {
+	err, output := runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() (error, string) {
 		return runTerraformWithRetry(terragruntOptions)
 	})
+	if err != nil && isErrorRequiringInit(output, err, terragruntOptions) {
+		initOptions := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+		initOptions.TerraformCommand = "init"
+		initOptions.TerraformCliArgs = []string{"init"}
+		initOptions.AutoInit = false
+		if err := runTerragruntWithConfig(initOptions, terragruntConfig, allowSourceDownload); err != nil {
+			return err
+		}
+
+		err, _ = runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() (error, string) {
+			return runTerraformWithRetry(terragruntOptions)
+		})
+	}
+
+	return err
 }
 
 // Run the given action function surrounded by hooks. That is, run the before hooks first, then, if there were no
 // errors, run the action, and finally, run the after hooks. Return any errors hit from the hooks or action.
-func runActionWithHooks(description string, terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig, action func() error) error {
+func runActionWithHooks(description string, terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig, action func() (error, string)) (error, string) {
 	beforeHookErrors := processHooks(terragruntConfig.Terraform.GetBeforeHooks(), terragruntOptions)
 
-	var actionErrors error
+	var (
+		actionErrors error
+		actionOutput string
+	)
 	if beforeHookErrors == nil {
-		actionErrors = action()
+		actionErrors, actionOutput = action()
```

Comment:
Does `actionOutput` contain `stdout`? Or `stderr`? Or both? Might want to make that clearer via the var name.

Reply:
Good point. It contains standard error output. I will update the name! Thank you!

Reply:
Done.